Bogus "Could not get the source for ..."

The problem manifests itself e.g. with specifying non-existent patch name in opam file (#986). Consider : 

```
=-=-= Installing mikmatch.1.0.6+patch =-=-=
mikmatch.1.0.6+patch Downloading http://mjambon.com/releases/mikmatch/mikmatch-1.0.6.tar.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 64798  100 64798    0     0  30928      0  0:00:02  0:00:02 --:--:-- 30944
200
Copying ~/.opam/repo/my/packages/mikmatch.1.0.6+patch/files/add_subpackages.patch to ~/.opam/4.00.1/build/mikmatch.1.0.6+patch/
Applying add-subpackages.patch.

[ERROR] Due to some errors while processing mikmatch.1.0.6+my, the following actions will NOT proceed:
 - install myproject.git

===== ERROR while installing mikmatch.1.0.6+patch =====
Could not get the source for mikmatch.1.0.6+patch.
```

When looking carefully one can see that the names of the patch file copied and applied are different. But the error message is completely bogus and misleading.
The problem is in silent catching of exceptions which is by any means unacceptable especially in the project of such scale. I don't know the opam build-system so cannot check why "unused value" warning is disabled in this case. Quick fix : 

```
diff --git a/src/client/opamAction.ml b/src/client/opamAction.ml
index 61e6838..21f6b3f 100644
--- a/src/client/opamAction.ml
+++ b/src/client/opamAction.ml
@@ -540,7 +540,7 @@ let build_and_install_package_aux t ~metadata nv =
           OpamFilename.exec ~env ~name ~metadata p_build commands in

       Some exec
-    with e -> None
+    with e -> OpamGlobals.msg "error: %s" (Printexc.to_string e); None
   in
   match exec with
   | None ->
```

Of course the real fix is to enable the warning and ensure that no such code slips through in the future.

--- a/src/client/opamAction.ml
+++ b/src/client/opamAction.ml
@@ -540,16 +540,14 @@ let build_and_install_package_aux t ~metadata nv =
           let metadata = get_metadata t in
           OpamFilename.exec ~env ~name ~metadata p_build commands in
 
-      Some exec
-    with e -> None
+      exec
+    with e ->
+      raise
+        (OpamGlobals.Package_error
+           (Printf.sprintf "Could not get the source for %s:\n%s"
+              (OpamPackage.to_string nv)
+              (Printexc.to_string e)))
   in
-  match exec with
-  | None ->
-    raise
-      (OpamGlobals.Package_error
-         (Printf.sprintf "Could not get the source for %s."
-            (OpamPackage.to_string nv)))
-  | Some exec ->
     try
       (* First, we build the package. *)
       exec ("Building " ^ OpamPackage.to_string nv) OpamFile.OPAM.build;

--- a/src/client/opamClient.ml
+++ b/src/client/opamClient.ml
@@ -150,7 +150,7 @@ let with_switch_backup command f =
        OpamPackage.Set.equal t.installed_roots t1.installed_roots then
       OpamFilename.remove file
     else
-      Printf.eprintf "The former state can be restored with \
+      Printf.eprintf "\nThe former state can be restored with \
                       %s switch import -f %S\n%!"
         Sys.argv.(0) (OpamFilename.to_string file);
     raise err

--- a/src/core/opamSystem.ml
+++ b/src/core/opamSystem.ml
@@ -591,9 +591,9 @@ let download ~overwrite ~filename:src ~dst:dst =
     really_download ~overwrite ~src ~dst
 
 let patch p =
-  let max_trying = 20 in
+  let max_trying = 4 in
   if not (Sys.file_exists p) then
-    internal_error "Cannot find %s." p;
+    internal_error "Patch file %S not found." p;
   let patch ~dryrun n =
     let opts = if dryrun then
         let open OpamGlobals in
@@ -607,7 +607,7 @@ let patch p =
     command ?verbose ("patch" :: ("-p" ^ string_of_int n) :: "-i" :: p :: opts) in
   let rec aux n =
     if n = max_trying then
-      internal_error "Application of %s failed: can not determine the correct patch level." p
+      internal_error "Patch %s does not apply." p
     else if None = try Some (patch ~dryrun:true n) with _ -> None then
       aux (succ n)
     else
@@ -633,6 +633,7 @@ let () =
     | Process_error r     -> Some (OpamProcess.string_of_result r)
     | Internal_error m    -> Some (with_opam_info m)
     | Command_not_found c -> Some (Printf.sprintf "%S: command not found." c)
+    | Sys.Break           -> Some "User interruption"
     | Unix.Unix_error (e, fn, msg) ->
       let msg = if msg = "" then "" else " on " ^ msg in
       let error = Printf.sprintf "%s: %S failed%s: %s"

